### ISSUE TITLE
Issue accessing page extension on ancestors

### ISSUE THREAD

**Issue by marksweb** [divio/django-cms#3585]
I've got a simple page extension designed to associate a colour with a page in order to style the page accordingly. I'm having trouble figuring out how to get page attributes in a django shop category page as it's not associated with the CMS page. You get to the category page from a CMS page, so could the referring page attribute be passed in the view context?

    class PageColourExtension(PageExtension):
        """
        Extend the page model with a background image field.
    
        This would be better suited outside of section_blocks as it's more generic
        but shares the same colours as section_blocks so it's here for now.
        """
        page_colour = models.CharField(
            _("Base colour"),
            max_length=50,
            null=True,
            choices=COLOUR_CHOICES,
            help_text=_(
                "The colour which this page and it's children will be styled with"
            )
        )
        objects = PageColourManager()
        cache = CachedPageColourManager(fields=['extended_object_id'])
    
        def __unicode__(self):
            """
            Identify by the page it is tied to.
    
            @return: page name
            @rtype: str
            """
            return '%s' % self.extended_object
    
    
    extension_pool.register(PageColourExtension)

The structure of the site is based on a number of sections containing multiple child pages. For simplicity I've attempted to access the top level ancestor to get that page colour on any child pages so my base template contains the following;

    {% block base-section-element %}
        {% if request.current_page.get_cached_ancestors|length <= 1 %}
            {% cache 900 base-section-element request.current_page.pagecolourextension.page_colour LANGUAGE_CODE %}
                <section id="content"
                         class="{{ request.current_page.pagecolourextension.page_colour }}">
            {% endcache %}
        {% else %}
            {% for ance in request.current_page.get_cached_ancestors %}
                {% if ance.level == 1 %}
                    {% cache 900 base-section-element ance.pagecolourextension.page_colour LANGUAGE_CODE %}
                        <section id="content"
                                 class="{{ ance.pagecolourextension.page_colour }} {{ request.current_page.pagecolourextension.page_colour }}">
                    {% endcache %}
                {% endif %}
            {% endfor %}
        {% endif %}
    {% endblock base-section-element %}

I'm debugging `get_cached_ancestors()` and can see that in the ancestors my `pagecolourextension` attribute returns an error;

	Traceback (most recent call last):
	  File "/Applications/PyCharm.app/helpers/pydev/pydevd_resolver.py", line 187, in _getPyDictionary
	    attr = getattr(var, n)
	  File "/Users/mwalker/Sites/vmlm/lib/python2.7/site-packages/django/db/models/fields/related.py", line 206, in __get__
	    self.related.get_accessor_name()))
	DoesNotExist: Page has no pagecolourextension.

But the top level page that I can see does have a `pagecolourextension` object associated with it so I don't know if it is because `get_cached_ancestors()` is using cache and therefore hasn't been invalidated when I saved the page.

Any ideas guys? Its been a few months since I developed this so I've lost touch with CMS page extensions.

**Comment by mkoistinen** [comment]
@marksweb interesting issue. I think we need to perform some troubleshooting for this one at your end. Looks like you're caching stuff for 15 minutes, it may be possible that you posted this issue during that period, and as you point out above, it could have been that the value just plain didn't exist when it was cached. Does this issue still exist?

While there may indeed be some form of issue, I would still suggest that you check for the existence of the pagecolourextension before you attempt to access properties off it. Being defensive here will help you while we find the issue.

Can you try this in 3.0.10, and let us know if the issue persists? Thanks.

**Comment by marksweb** [comment]
Thanks @mkoistinen I'll take a look at it when I get chance next week. I've changed things somewhat since I posted this, removed the majority of template caching, improved my instance caching & added this feature to my menus.

I've just updated my sites to 3.0.10 so I'll take another look at this - currently I've still got hard coded colours in the template for the django shop pages I believe.
And to take this opportunity for a little name (URL) drop... http://www.virginmoneylondonmarathon.com/en-gb/shopping/shop/
